File: lab-06/scripts/lab-0606-analysis/Comparison.scala
import scala.util.Try

object Comparison {
  // Модель данных
  case class User(id: Int, name: String, email: String)
  case class Product(id: Int, name: String, price: Double, category: String)
  case class OrderItem(product: Product, quantity: Int)
  case class Order(id: Int, user: User, items: List[OrderItem], status: String)
  
  // Пример данных
  val users = List(
    User(1, "John Doe", "john@example.com"),
    User(2, "Jane Smith", "jane@example.com")
  )
  
  val products = List(
    Product(1, "iPhone", 999.99, "electronics"),
    Product(2, "MacBook", 1999.99, "electronics"),
    Product(3, "T-shirt", 29.99, "clothing")
  )
  
  val orders = List(
    Order(1, users(0), List(OrderItem(products(0), 1), OrderItem(products(2), 2)), "completed"),
    Order(2, users(1), List(OrderItem(products(1), 1)), "pending")
  )
  
  // Функции обработки
  def calculateOrderTotal(order: Order): Double = 
    order.items.map(item => item.product.price * item.quantity).sum
  
  def filterOrdersByStatus(orders: List[Order], status: String): List[Order] = 
    orders.filter(_.status == status)
  
  def getTopExpensiveOrders(orders: List[Order], n: Int): List[Order] = 
    orders.sortBy(calculateOrderTotal)(Ordering[Double].reverse).take(n)
  
  def applyDiscount(order: Order, discount: Double): Order = {
    val discountedItems = order.items.map { item =>
      item.copy(product = item.product.copy(price = item.product.price * (1 - discount)))
    }
    order.copy(items = discountedItems)
  }
  
  // ===== ✓ ПРАКТИЧЕСКИЕ ЗАДАНИЯ =====
  // Задание 3: Сравнение обработки ошибок (ТОЧНО ИЗ УСЛОВИЯ)
  def safeCalculateTotal(order: Order): Option[Double] = {
    Try {
      order.items.map(item => item.product.price * item.quantity).sum
    }.toOption
  }
  
  def main(args: Array[String]): Unit = {
    println(s"Scala ✓ safeCalculateTotal: ${safeCalculateTotal(orders.head)}")  // ✓ Задание 3
    
    val completedOrders = filterOrdersByStatus(orders, "completed")
    val totalRevenue = completedOrders.map(calculateOrderTotal).sum
    val topOrders = getTopExpensiveOrders(orders, 2)
    
    println(s"Общая выручка: $totalRevenue")
    println(s"Топ заказы: ${topOrders.map(calculateOrderTotal)}")
  }
}

File: lab-06/scripts/lab-0606-analysis/comparison.py
from dataclasses import dataclass
from typing import List, Dict
import timeit

@dataclass
class User:
    id: int
    name: str
    email: str

@dataclass
class Product:
    id: int
    name: str
    price: float
    category: str

@dataclass
class OrderItem:
    product: Product
    quantity: int

@dataclass
class Order:
    id: int
    user: User
    items: List[OrderItem]
    status: str

# Пример данных
users = [
    User(1, "John Doe", "john@example.com"),
    User(2, "Jane Smith", "jane@example.com")
]

products = [
    Product(1, "iPhone", 999.99, "electronics"),
    Product(2, "MacBook", 1999.99, "electronics"),
    Product(3, "T-shirt", 29.99, "clothing")
]

orders = [
    Order(1, users[0], [OrderItem(products[0], 1), OrderItem(products[2], 2)], "completed"),
    Order(2, users[1], [OrderItem(products[1], 1)], "pending")
]

# Функции обработки
def calculate_order_total(order: Order) -> float:
    return sum(item.product.price * item.quantity for item in order.items)

def filter_orders_by_status(orders: List[Order], status: str) -> List[Order]:
    return list(filter(lambda order: order.status == status, orders))

def get_top_expensive_orders(orders: List[Order], n: int) -> List[Order]:
    return sorted(orders, key=calculate_order_total, reverse=True)[:n]

def apply_discount(order: Order, discount: float) -> Order:
    discounted_items = [
        OrderItem(
            Product(
                item.product.id,
                item.product.name,
                item.product.price * (1 - discount),
                item.product.category
            ),
            item.quantity
        ) for item in order.items
    ]
    return Order(order.id, order.user, discounted_items, order.status)

# ===== ✓ ПРАКТИЧЕСКИЕ ЗАДАНИЯ =====
# Задание 1: Измерение производительности (ТОЧНО ИЗ УСЛОВИЯ)
def benchmark_python():
    return timeit.timeit(lambda: sum(calculate_order_total(order) for order in orders), number=1000)

def main():
    print(f"Python ✓ Бенчмарк: {benchmark_python()*1000:.2f}ms")  # ✓ Задание 1
    
    completed_orders = filter_orders_by_status(orders, "completed")
    total_revenue = sum(calculate_order_total(order) for order in completed_orders)
    top_orders = get_top_expensive_orders(orders, 2)
    
    print(f"Общая выручка: {total_revenue}")
    print(f"Топ заказы: {[calculate_order_total(order) for order in top_orders]}")

if __name__ == "__main__":
    main()
